### PR TITLE
Implement lazy member loading for mirrored protocol members in imported types

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@jrose-apple The second patch doesn't have the expected performance gain because while we spend less time in loadAllMembers(), the lookupDirect() path has slowed down because a lazy member 'miss' does so much extra work. So I need to optimize something there. But comments on the general approach are welcome. I might land the first commit separately if all else fails, although it would be a shame.

**Comment by jrose-apple** [comment]
This generally makes sense to me. I don't see that that much of the work is avoidable the way things are set up now, so maybe the real answer is going to have to be altering the lookup tables (which is not a bad idea anyway).

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Comment by jrose-apple** [comment]
Wow, we dropped **a third** of the Clang decls we imported. I can actually believe that. And yet we also regressed compile time by over 1%.

**Comment by CodaFi** [comment]
Can you strike the inherited constructors part of this patch?  It's a bit noisy.

**Review by CodaFi — APPROVED** [review]

File: lib/AST/NameLookup.cpp [hunk 1]
```diff
@@ -1143,37 +1143,30 @@ void ExtensionDecl::addedMember(Decl *member) {
 // MemberLookupTable is constructed (and possibly has entries in it),
 // MemberLookupTable is incrementally reconstituted with new members.
 
-static bool
+static void
 populateLookupTableEntryFromLazyIDCLoader(ASTContext &ctx,
                                           MemberLookupTable &LookupTable,
                                           DeclBaseName name,
                                           IterableDeclContext *IDC) {
+  assert(!IDC->isLoadingLazyMembers());
```
Comment (CodaFi) [inline lib/AST/NameLookup.cpp]: Let's see if your assert can stand up to the bots.  You might be in the same position I am right now with the lookup requests.

**Comment by slavapestov** [comment]
@CodaFi The inherited constructors change is in a separate NFC commit.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test compiler performance

**Comment by CodaFi** [comment]
@swift-ci please smoke test compiler performance

**Pull request merged**